compiler/src/dmd/dtemplate.d

```diff
@@ -3843,62 +3843,63 @@ MATCH deduceType(RootObject o, Scope* sc, Type tparam, TemplateParameters* param
+                    assert(i != IDX_NOTFOUND);
+                    tp = (*parameters)[i];
                 }
-                else if (tparam.ty == Taarray)
+                else
+                    edim = tsa.dim;
+            }
+            else if (tparam.ty == Taarray)
+            {
+                TypeAArray taa = cast(TypeAArray)tparam;
+                i = templateParameterLookup(taa.index, parameters);
+                if (i != IDX_NOTFOUND)
+                    tp = (*parameters)[i];
+                else
                 {
-                    TypeAArray taa = cast(TypeAArray)tparam;
-                    i = templateParameterLookup(taa.index, parameters);
-                    if (i != IDX_NOTFOUND)
-                        tp = (*parameters)[i];
-                    else
+                    Loc loc;
+                    // The "type" (it hasn't been resolved yet) of the function parameter
+                    // does not have a location but the parameter it is related to does,
+                    // so we use that for the resolution (better error message).
+                    if (inferStart < parameters.length)
                     {
-                        Loc loc;
-                        // The "type" (it hasn't been resolved yet) of the function parameter
-                        // does not have a location but the parameter it is related to does,
-                        // so we use that for the resolution (better error message).
-                        if (inferStart < parameters.length)
-                        {
-                            TemplateParameter loctp = (*parameters)[inferStart];
-                            loc = loctp.loc;
-                        }
-
-                        Expression e;
-                        Type tx;
-                        Dsymbol s;
-                        taa.index.resolve(loc, sc, e, tx, s);
-                        edim = s ? getValue(s) : getValue(e);
+                        TemplateParameter loctp = (*parameters)[inferStart];
+                        loc = loctp.loc;
                     }
+
+                    Expression e;
+                    Type tx;
+                    Dsymbol s;
+                    taa.index.resolve(loc, sc, e, tx, s);
+                    edim = s ? getValue(s) : getValue(e);
                 }
-                if (tp && tp.matchArg(sc, t.dim, i, parameters, dedtypes, null) || edim && edim.toInteger() == t.dim.toInteger())
-                {
-                    result = deduceType(t.next, sc, tparam.nextOf(), parameters, dedtypes, wm);
-                    return;
-                }
             }
+            if (tp && tp.matchArg(sc, t.dim, i, parameters, dedtypes, null) || edim && edim.toInteger() == t.dim.toInteger())
+            {
+                result = deduceType(t.next, sc, tparam.nextOf(), parameters, dedtypes, wm);
+                return;
+            }
+
             visit(cast(Type)t);
```

Comment:
You moved the call to this function to the `if (!tparam)` case, so shouldn't this be deleted here?
```suggestion
```